docs: CI Integrations Config Validator design document (#990)
platform-platform/monorepo: metrics/ci_integrations/docs/04_ci_integrations_config_validator.md

# CI Integrations Config Validator

## TL;DR

Introducing a `CI Integrations Config Validator` provides an ability to validate the CI Integrations configuration file before the `sync` command to provide additional context about the possible errors in the configuration file.  
For example, if the configuration file contains a non-valid email/password used to log in into CI, the user sees the corresponding error before the `sync` process starts.

## References
> Link to supporting documentation, GitHub tickets, etc.

- [CI Integrations Tool Architecture](https://github.com/platform-platform/monorepo/blob/master/metrics/ci_integrations/docs/01_ci_integration_module_architecture.md)

## Goals
> Identify success metrics and measurable goals.

This document aims the following goals: 
- Create a clear design of the CI Integrations Config Validator.
- Provide an overview of steps the new CI Integrations Config Validator requires.

## Design
> Explain and diagram the technical design.

Consider the following steps needed to be able to validate the given configuration file:

> 1. Create abstract `ConfigValidator`, `ConfigValidatorClient`, and `ConfigValidatorFactory` classes.

Reviewer: Does it make sense to explain the main responsibilities of these classes?

- [CI Integrations Tool Architecture](https://github.com/platform-platform/monorepo/blob/master/metrics/ci_integrations/docs/01_ci_integration_module_architecture.md)

## Goals
> Identify success metrics and measurable goals.

This document aims the following goals: 
- Create a clear design of the CI Integrations Config Validator.
- Provide an overview of steps the new CI Integrations Config Validator requires.

## Design
> Explain and diagram the technical design.

Let's start with the necessary abstractions. Consider the following classes:
- A `ConfigValidator` is a class that provides the validation functionality and throws a `ConfigValidationException` if the given config is not valid.
- A `ValidationDelegate` is a class that the `ConfigValidator` uses for the validation of specific fields with network calls.
- A `ConfigValidatorFactory` is a class that creates a `ConfigValidator` with its `ValidationDelegate`.

Consider the following steps needed to be able to validate the given configuration file:

1. Create the following abstract classes: `ConfigValidator`, `ValidationDelegate`, `SourceValidationDelegate`, `DestinationValidationDelegate` and `ConfigValidatorFactory`.

> 2. For each source or destination party implement its specific `ConfigValidator`, `ValidationDelegate`, and `ConfigValidatorFactory`. Provide integration-specific clients with the validation-required methods.

Reviewer: ```suggestion
2. For each source or destination party, implement its specific `ConfigValidator`, `ValidationDelegate`, and `ConfigValidatorFactory`. Implement the validation-required methods in the integration-specific clients.
```
?
